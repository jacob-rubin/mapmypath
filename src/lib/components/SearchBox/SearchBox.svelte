<script lang="ts">
	import Autocomplete from './Autocomplete/Autocomplete.svelte';
	import Searcher from '$lib/utils/searcher/searcher.svelte';
	import { setSearcherContext } from './context/searcherContext';

	let isFocused = $state(false);

	async function handleFocus() {
		isFocused = true;
	}

	async function handleBlur() {
		isFocused = false;
	}

	const searcher: Searcher = new Searcher();
	setSearcherContext(searcher);
</script>

<div class="flex w-72 flex-col gap-2">
	<input
		bind:value={searcher.text}
		onfocus={handleFocus}
		onblur={handleBlur}
		data-testid="searchbox"
		type="text"
		placeholder="Search"
		class="input input-bordered"
		id="search"
	/>
	{#if isFocused}
		<Autocomplete suggestionResponse={searcher.suggestions}
		></Autocomplete>
	{/if}
</div>
