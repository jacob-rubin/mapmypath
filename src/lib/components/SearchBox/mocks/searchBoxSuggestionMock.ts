import type { SearchBoxSuggestion } from '@mapbox/search-js-core';

export const mockSearchSuggestionResponse: SearchBoxSuggestion = {
	name: 'Viking Park Iceland',
	mapbox_id:
		'dXJuOm1ieHBvaTpiY2ExMjRmYS0yZTlhLTQyN2EtYmM1NS05ZWY2NTMwNWJlMDI',
	feature_type: 'poi',
	address: 'Hjörleifshöfði',
	full_address: 'Hjörleifshöfði, 871 Vík í Mýrdal, Iceland',
	place_formatted: '871 Vík í Mýrdal, Iceland',
	context: {
		country: {
			id: '',
			name: 'Iceland',
			country_code: 'IS',
			country_code_alpha_3: 'ISL'
		},
		postcode: {
			id: 'dXJuOm1ieHBsYzpGWTV2',
			name: '871'
		},
		place: {
			id: 'dXJuOm1ieHBsYzpFSWh2',
			name: 'Vík í Mýrdal'
		},
		street: {
			id: '',
			name: 'hjörleifshöfði'
		}
	},
	language: 'en',
	maki: 'attraction',
	poi_category: ['tourist attraction'],
	external_ids: {
		dataplor: '108d8703-4f0f-40b8-8f51-d0f99c79d680'
	},
	metadata: {},
	name_preferred: '',
	brand: '',
	brand_id: '',
	distance: 0,
	eta: 0,
	added_distance: 0,
	added_time: 0
};
