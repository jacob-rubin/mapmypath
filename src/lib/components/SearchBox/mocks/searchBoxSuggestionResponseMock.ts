import type { SearchBoxSuggestionResponse } from '@mapbox/search-js-core';

export const mockSearchBoxSuggestionResponse: SearchBoxSuggestionResponse =
	{
		suggestions: [
			{
				name: 'Viking Park Iceland',
				mapbox_id:
					'dXJuOm1ieHBvaTpiY2ExMjRmYS0yZTlhLTQyN2EtYmM1NS05ZWY2NTMwNWJlMDI',
				feature_type: 'poi',
				address: 'Hjörleifshöfði',
				full_address: 'Hjörleifshöfði, 871 Vík í Mýrdal, Iceland',
				place_formatted: '871 Vík í Mýrdal, Iceland',
				context: {
					country: {
						id: '',
						name: 'Iceland',
						country_code: 'IS',
						country_code_alpha_3: 'ISL'
					},
					postcode: {
						id: 'dXJuOm1ieHBsYzpGWTV2',
						name: '871'
					},
					place: {
						id: 'dXJuOm1ieHBsYzpFSWh2',
						name: 'Vík í Mýrdal'
					},
					street: {
						id: '',
						name: 'hjörleifshöfði'
					}
				},
				language: 'en',
				maki: 'attraction',
				poi_category: ['tourist attraction'],
				external_ids: {
					dataplor: '108d8703-4f0f-40b8-8f51-d0f99c79d680'
				},
				metadata: {},
				name_preferred: '',
				brand: '',
				brand_id: '',
				distance: 0,
				eta: 0,
				added_distance: 0,
				added_time: 0
			},
			{
				name: 'Viking Journeys Iceland',
				mapbox_id:
					'dXJuOm1ieHBvaTpmZGYxZWRkNi05NGNhLTQ5Y2UtYWZjZS05M2YwNGNkMTkwOTc',
				feature_type: 'poi',
				address: 'Jörfabakki',
				full_address: 'Jörfabakki, 109 Reykjavík, Iceland',
				place_formatted: '109 Reykjavík, Iceland',
				context: {
					country: {
						id: '',
						name: 'Iceland',
						country_code: 'IS',
						country_code_alpha_3: 'ISL'
					},
					postcode: {
						id: 'dXJuOm1ieHBsYzpBUTV2',
						name: '109'
					},
					place: {
						id: 'dXJuOm1ieHBsYzpCaWh2',
						name: 'Reykjavík'
					},
					street: {
						name: 'jörfabakki',
						id: ''
					}
				},
				language: 'en',
				maki: 'marker',
				poi_category: ['services'],
				external_ids: {
					dataplor: '9bc1c213-7960-442d-b3e3-0882642135f2'
				},
				metadata: {},
				name_preferred: '',
				brand: '',
				brand_id: '',
				distance: 0,
				eta: 0,
				added_distance: 0,
				added_time: 0
			},
			{
				name: 'Vík',
				name_preferred: 'Vík í Mýrdal',
				mapbox_id: 'dXJuOm1ieHBsYzpFSWh2',
				feature_type: 'place',
				place_formatted: 'Southern Region, Iceland',
				context: {
					country: {
						id: 'dXJuOm1ieHBsYzpJbTg',
						name: 'Iceland',
						country_code: 'IS',
						country_code_alpha_3: 'ISL'
					},
					region: {
						id: 'dXJuOm1ieHBsYzpSRzg',
						name: 'Southern Region',
						region_code: '8',
						region_code_full: 'IS-8'
					}
				},
				language: 'en',
				maki: 'marker',
				metadata: {},
				address: '',
				full_address: '',
				poi_category: [],
				brand: '',
				brand_id: '',
				external_ids: undefined,
				distance: 0,
				eta: 0,
				added_distance: 0,
				added_time: 0
			},
			{
				name: 'Icelandic Mountain Guides',
				mapbox_id:
					'dXJuOm1ieHBvaTo0NTU0YzhjZC00YjRiLTQwMmQtYjBiZC03N2M4NjFjNWEwYzA',
				feature_type: 'poi',
				address: 'Icelandic Mountain Guides ATV&Snowmobile Tours',
				full_address:
					'Icelandic Mountain Guides ATV&Snowmobile Tours, 871 Vík í Mýrdal, Iceland',
				place_formatted: '871 Vík í Mýrdal, Iceland',
				context: {
					country: {
						id: '',
						name: 'Iceland',
						country_code: 'IS',
						country_code_alpha_3: 'ISL'
					},
					postcode: {
						id: 'dXJuOm1ieHBsYzpGWTV2',
						name: '871'
					},
					place: {
						id: 'dXJuOm1ieHBsYzpFSWh2',
						name: 'Vík í Mýrdal'
					}
				},
				language: 'en',
				maki: 'attraction',
				poi_category: ['tourist attraction', 'tours'],
				external_ids: {
					dataplor: 'f8a2c0e2-436e-4ffb-a467-9b448efb8003'
				},
				metadata: {},
				name_preferred: '',
				brand: '',
				brand_id: '',
				distance: 0,
				eta: 0,
				added_distance: 0,
				added_time: 0
			},
			{
				name: 'Icelandic Mountain Guides',
				mapbox_id:
					'dXJuOm1ieHBvaTo0Mjg2NTQ0ZC1kMWU1LTQzZGQtODg2Ni0wNzY3YjcxN2NhOTk',
				feature_type: 'poi',
				address: '871 Vík í Mýrdal',
				full_address: '871 Vík í Mýrdal, Iceland',
				place_formatted: 'Iceland',
				context: {
					country: {
						id: '',
						name: 'Iceland',
						country_code: 'IS',
						country_code_alpha_3: 'ISL'
					},
					postcode: {
						id: 'dXJuOm1ieHBsYzpGWTV2',
						name: '871'
					},
					place: {
						id: 'dXJuOm1ieHBsYzpFSWh2',
						name: 'Vík í Mýrdal'
					}
				},
				language: 'en',
				maki: 'attraction',
				poi_category: ['tourist attraction', 'tours'],
				external_ids: {
					dataplor: 'cca72ee9-820d-4c54-9480-8d21cf1838e5'
				},
				metadata: {},
				name_preferred: '',
				brand: '',
				brand_id: '',
				distance: 0,
				eta: 0,
				added_distance: 0,
				added_time: 0
			}
		],
		attribution:
			'© 2025 Mapbox and its suppliers. All rights reserved. Use of this data is subject to the Mapbox Terms of Service. (https://www.mapbox.com/about/maps/)',
		url: 'https://api.mapbox.com/search/searchbox/v1/suggest?q=Vik%2C%20Iceland&access_token=pk.eyJ1IjoiamFjcnViMTgiLCJhIjoiY201Z3hld2gwMDVtbzJqb29jaHBncmhjbyJ9.AXOC8KRomVWR1FYelUcptA&language=en&limit=5&session_token=085e0dcd-f441-4861-8944-cb8cafa0fb59.0'
	};
