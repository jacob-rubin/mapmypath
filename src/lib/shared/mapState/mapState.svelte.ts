import type { MarkerData } from '$lib/types/markerData';
import type { LngLat } from 'mapbox-gl';

class MapState {
	#markers: MarkerData[] = $state([]); //TODO: consider making this a hash map

	constructor() {
		this.#markers = [];
	}

	addMarker(marker: MarkerData) {
		this.#markers.push(marker);
		console.log($state.snapshot(this.#markers));
	}

	updateMarker(id: number | string, lngLat: LngLat) {
		const marker = this.#markers.find((marker) => marker.id === id);
		if (!marker) {
			throw new Error(`Marker with id ${id} not found`);
		}

		marker.lngLat = lngLat;
	}

	getMarkers() {
		return this.#markers;
	}

	getMarker(id: number): MarkerData {
		const marker: MarkerData | undefined = this.#markers.find(
			(marker) => marker.id === id
		);

		if (!marker) {
			throw new Error(`Marker with id ${id} not found`);
		}

		return marker;
	}

	clear() {
		this.#markers = [];
	}
}

export const mapState = new MapState();
