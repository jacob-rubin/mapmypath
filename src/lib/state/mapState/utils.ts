import type { MapState } from '$lib/state/mapState/mapState.svelte';
import type Mapbox from '$lib/utils/mapbox/mapbox';
import Marker, {
	type MarkerData
} from '$lib/utils/marker/marker.svelte';

export function addMapClickListeners(
	map: Mapbox,
	mapState: MapState
) {
	map.addClickListener((e: mapboxgl.MapMouseEvent) => {
		mapState.addMarker(e.lngLat);
	});
}

// TODO: Clean this up, since don't like having function with 3 inputs
export function addMarkerDragListeners(
	marker: Marker,
	mapState: MapState,
	map: Mapbox
) {
	marker.addDragListener((markerData: MarkerData) => {
		mapState.updateMarker({
			id: markerData.id,
			lngLat: markerData.lngLat
		});
		map.renderPath(mapState.markers.map((m) => m.lngLat));
	});
}
