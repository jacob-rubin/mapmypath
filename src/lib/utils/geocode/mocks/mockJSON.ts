export const mockGeocodeFailJSON = {
	type: 'FeatureCollection',
	features: [],
	attribution:
		'NOTICE: © 2024 Mapbox and its suppliers. All rights reserved. Use of this data is subject to the Mapbox Terms of Service (https://www.mapbox.com/about/maps/). This response and the information it contains may not be retained.'
};

export const mockGeocodeSuccessJSON = {
	type: 'FeatureCollection',
	features: [
		{
			type: 'Feature',
			id: 'dXJuOm1ieGFkcjoyODE5ZGJmNC1kOTg5LTQwODAtYWM5NC1hNDE4ZWQ0NzFjYWQ',
			geometry: {
				type: 'Point',
				coordinates: [-74.013791, 40.801782]
			},
			properties: {
				mapbox_id:
					'dXJuOm1ieGFkcjoyODE5ZGJmNC1kOTg5LTQwODAtYWM5NC1hNDE4ZWQ0NzFjYWQ',
				feature_type: 'address',
				full_address:
					'7417 John F Kennedy Boulevard, North Bergen, New Jersey 07047, United States',
				name: '7417 John F Kennedy Boulevard',
				name_preferred: '7417 John F Kennedy Boulevard',
				coordinates: {
					longitude: -74.013791,
					latitude: 40.801782,
					accuracy: 'rooftop',
					routable_points: [
						{
							name: 'default',
							latitude: 40.80159,
							longitude: -74.013262
						}
					]
				},
				place_formatted:
					'North Bergen, New Jersey 07047, United States',
				context: {
					address: {
						mapbox_id:
							'dXJuOm1ieGFkcjoyODE5ZGJmNC1kOTg5LTQwODAtYWM5NC1hNDE4ZWQ0NzFjYWQ',
						address_number: '7417',
						street_name: 'John F Kennedy Boulevard',
						name: '7417 John F Kennedy Boulevard'
					},
					street: {
						mapbox_id:
							'dXJuOm1ieGFkci1zdHI6MjgxOWRiZjQtZDk4OS00MDgwLWFjOTQtYTQxOGVkNDcxY2Fk',
						name: 'John F Kennedy Boulevard'
					},
					postcode: {
						mapbox_id: 'postcode.4022711030774558',
						name: '07047'
					},
					place: {
						mapbox_id: 'dXJuOm1ieHBsYzpEZzlJN0E',
						name: 'North Bergen',
						wikidata_id: 'Q2000528',
						translations: {
							en: { language: 'en', name: 'North Bergen' }
						}
					},
					district: {
						mapbox_id: 'dXJuOm1ieHBsYzpwQWJz',
						name: 'Hudson County',
						wikidata_id: 'Q490505',
						translations: {
							en: { language: 'en', name: 'Hudson County' }
						}
					},
					region: {
						mapbox_id: 'dXJuOm1ieHBsYzpBbVRz',
						name: 'New Jersey',
						wikidata_id: 'Q1408',
						region_code: 'NJ',
						region_code_full: 'US-NJ',
						translations: {
							en: { language: 'en', name: 'New Jersey' }
						}
					},
					country: {
						mapbox_id: 'dXJuOm1ieHBsYzpJdXc',
						name: 'United States',
						wikidata_id: 'Q30',
						country_code: 'US',
						country_code_alpha_3: 'USA',
						translations: {
							en: { language: 'en', name: 'United States' }
						}
					}
				}
			}
		}
	],
	attribution:
		'NOTICE: © 2024 Mapbox and its suppliers. All rights reserved. Use of this data is subject to the Mapbox Terms of Service (https://www.mapbox.com/about/maps/). This response and the information it contains may not be retained.'
};
