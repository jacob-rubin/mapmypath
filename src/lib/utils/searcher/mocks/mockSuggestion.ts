import type { SuggestionResponse } from '../types/types';

export const mockSuggestion: SuggestionResponse = {
	suggestions: [
		{
			name: 'Michigan Stadium',
			mapbox_id:
				'dXJuOm1ieHBvaTpjYjUyYTZkYy03ZjNkLTRhZjctYTJhMi02NTIyMzk3YjczZmM',
			feature_type: 'poi',
			address: '1201 S Main St',
			full_address:
				'1201 S Main St, Ann Arbor, Michigan 48104, United States',
			place_formatted: 'Ann Arbor, Michigan 48104, United States',
			context: {
				country: {
					name: 'United States',
					country_code: 'US',
					country_code_alpha_3: 'USA'
				},
				region: {
					name: 'Michigan',
					region_code: 'MI',
					region_code_full: 'US-MI'
				},
				postcode: { id: 'dXJuOm1ieHBsYzpDYlpPN0E', name: '48104' },
				place: { id: 'dXJuOm1ieHBsYzpoQ2pz', name: 'Ann Arbor' },
				neighborhood: {
					id: 'dXJuOm1ieHBsYzpJODBNN0E',
					name: 'South Main'
				},
				address: {
					name: '1201 S Main St',
					address_number: '1201',
					street_name: 's main st'
				},
				street: { name: 's main st' }
			},
			language: 'en',
			maki: 'marker',
			poi_category: ['sports', 'stadium'],
			poi_category_ids: ['sports', 'stadium'],
			external_ids: {
				dataplor: '1e086643-5035-430f-875e-e6eaedc6dbd6'
			},
			metadata: {},
			operational_status: 'active'
		}
	],
	attribution:
		'Â© 2023 Mapbox and its suppliers. All rights reserved. Use of this data is subject to the Mapbox Terms of Service. (https://www.mapbox.com/about/maps/)'
};
