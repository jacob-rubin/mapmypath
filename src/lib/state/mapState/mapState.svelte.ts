import { addMapClickListeners } from '$lib/state/mapState/utils';
import type Mapbox from '$lib/utils/mapbox/mapbox';
import Marker from '$lib/utils/marker/marker.svelte';
import type { MarkerData } from '$lib/utils/marker/marker.svelte';

export class MapState {
	#map: Mapbox;
	#markers: Marker[] = $state([]); //TODO: consider making this collection instead of a list

	#counter: number = 0; // TODO: Extrapolate this into a counter class?

	constructor(map: Mapbox) {
		this.#map = map;
		this.#markers = [];

		addMapClickListeners(this.#map, this);
	}

	get markers(): Marker[] {
		return this.#markers;
	}

	set markers(value: Marker[]) {
		this.#markers = value;
	}

	get map(): Mapbox {
		return this.#map;
	}

	getMarkerById(id: number): Marker {
		const marker: Marker | undefined = this.#markers.find(
			(marker) => marker.id === id
		);

		if (!marker) {
			throw new Error(`Marker with id ${id} not found`);
		}

		return marker;
	}

	addMarker(lngLat: mapboxgl.LngLat): Marker {
		// TODO: How can I make this method smaller?
		const marker: Marker = new Marker({
			id: this.#counter++,
			lngLat,
			name: `Stop ${this.#markers.length + 1}`
		});

		marker.addDragListener((markerData: MarkerData) => {
			this.updateMarker({
				id: markerData.id,
				lngLat: markerData.lngLat
			});
			this.#map.renderPath(this.#markers.map((m) => m.lngLat));
		});

		this.#markers.push(marker);
		this.#map.addMarker(marker);
		this.#map.renderPath(this.#markers.map((m) => m.lngLat));

		return marker;
	}

	updateMarker(markerData: MarkerData): void {
		// TODO: Do I need to update the marker lnglat, since it's only updated when dragged?
		//TODO: Can I remove this usage of MarkerData and replace with marker?
		const marker: Marker | undefined = this.#markers.find(
			(marker) => marker.id === markerData.id
		);

		if (!marker) {
			throw new Error(`Marker with id ${markerData.id} not found`);
		}

		marker.lngLat = markerData.lngLat;
	}

	deleteMarker(id: number): Marker {
		const marker: Marker | undefined = this.#markers.find(
			(marker) => marker.id === id
		);

		if (!marker) {
			throw new Error(`Marker with id ${id} not found`);
		}

		this.#markers = this.#markers.filter(
			(marker) => marker.id !== id
		);
		this.#map.deleteMarker(marker);
		this.#map.renderPath(this.#markers.map((m) => m.lngLat));

		return marker;
	}
}
