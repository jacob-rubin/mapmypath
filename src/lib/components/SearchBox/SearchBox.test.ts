import {
	afterEach,
	beforeEach,
	describe,
	expect,
	it,
	vi
} from 'vitest';
import SearchBox from './SearchBox.svelte';
import {
	cleanup,
	queryByPlaceholderText,
	render,
	within
} from '@testing-library/svelte';
import userEvent from '@testing-library/user-event';
import { mockSuggestion } from '$lib/utils/searcher/mocks/mockSuggestion';
import { tick } from 'svelte';
import SearchboxFixture from './fixtures/SearchboxFixture.svelte';
import Mapbox from '$lib/utils/mapbox/mapbox';
import { MapState } from '$lib/state/mapState/mapState.svelte';
import { mockFeature } from '$lib/utils/searcher/mocks/mockFeature';

function renderMap(): Mapbox {
	const mapElement: HTMLElement = document.createElement('div');
	mapElement.innerHTML = `<div id="map" class="h-80 w-80"></div`;
	document.body.appendChild(mapElement);

	const map: Mapbox = new Mapbox(mapElement);
	return map;
}

describe('SearchBox', async () => {
	let map: Mapbox;
	let mapState: MapState;

	beforeEach(async () => {
		map = renderMap();
		mapState = new MapState(map);
		await map.awaitLoad();
		await map.initializeStyles();
	});

	afterEach(() => {
		cleanup();
	});

	it('matches the snapshot', async ({ expect }) => {
		const screen = render(SearchBox);

		expect(screen.container.innerHTML).toMatchSnapshot();
	});

	it('has the placeholder `Search`', async () => {
		const screen = render(SearchBox);
		const input: HTMLElement = screen.getByRole('textbox');
		const placeholder: Element | null = queryByPlaceholderText(
			input,
			'Search'
		);

		expect(placeholder).toBeDefined();
	});

	it('does not display the autocomplete when the search text is empty', async () => {
		const screen = render(SearchBox);

		expect(screen.getByRole('textbox')).toHaveValue('');
		expect(screen.queryByTestId('autocomplete')).toBeNull();
	});

	it('sends a fetch request when typing', async () => {
		const mockFetch = vi.fn().mockResolvedValue(
			new Response(JSON.stringify(mockSuggestion), {
				status: 200,
				headers: { 'Content-Type': 'application/json' }
			})
		);
		vi.stubGlobal('fetch', mockFetch);

		const user = userEvent.setup();
		const screen = render(SearchBox);
		const textbox: HTMLElement = screen.getByRole('textbox');

		await user.type(textbox, 'Michigan Stadium');
		await new Promise((resolve) => setTimeout(resolve, 200));

		const fetchCall: string =
			mockFetch.mock.calls[0][0].split('&')[0];

		expect(mockFetch).toHaveBeenCalledOnce();
		expect(fetchCall).toBe(
			'https://api.mapbox.com/search/searchbox/v1/suggest?q=Michigan+Stadium'
		);

		vi.unstubAllGlobals();
	});

	it('hides the autocomplete when searchbox not focused', async () => {
		const user = userEvent.setup();
		const mockFetch = vi.fn().mockResolvedValue({
			json: vi.fn().mockResolvedValue(mockSuggestion),
			ok: true,
			status: 200,
			headers: { 'Content-Type': 'application/json' }
		});
		vi.stubGlobal('fetch', mockFetch);

		const screen = render(SearchBox);
		const textbox: HTMLElement = screen.getByRole('textbox');
		await user.type(textbox, 'Michigan Stadium');

		expect(screen.getByTestId('autocomplete')).toBeDefined();
		textbox.blur();
		await tick();
		expect(
			screen.queryByTestId('autocomplete')
		).not.toBeInTheDocument();
	});

	it('pastes the suggestion name into the searchbox when clicked', async () => {
		const user = userEvent.setup();
		const mockSuggestFetch = vi.fn().mockResolvedValue({
			json: vi.fn().mockResolvedValue(mockSuggestion),
			ok: true,
			status: 200,
			headers: { 'Content-Type': 'application/json' }
		});
		vi.stubGlobal('fetch', mockSuggestFetch);

		const screen = render(SearchboxFixture, {
			props: {
				context: {
					mapState: mapState
				}
			}
		});
		const textbox: HTMLElement = screen.getByRole('textbox');
		await user.type(textbox, 'Mich');

		const autocomplete = screen.getByTestId('autocomplete');
		await new Promise((resolve) => setTimeout(resolve, 200)); // Await debounce
		const suggestion = within(autocomplete).getByRole('button');

		vi.unstubAllGlobals();

		// add a mock feature for feature
		const mockFeatureFetch = vi.fn().mockResolvedValue({
			json: vi.fn().mockResolvedValue(mockFeature),
			ok: true,
			status: 200,
			headers: { 'Content-Type': 'application/json' }
		});
		vi.stubGlobal('fetch', mockFeatureFetch);

		await user.click(suggestion);
		await tick();
		expect(textbox).toHaveValue('Michigan Stadium');

		vi.unstubAllGlobals();
	});
});
