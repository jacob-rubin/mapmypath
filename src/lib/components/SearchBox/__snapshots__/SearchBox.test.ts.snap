// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBox > matches the snapshot 1`] = `"<div class="flex w-72 flex-col gap-2"><input data-testid="searchbox" type="text" placeholder="Search" class="input input-bordered" id="search"> <!----></div>"`;
